module/private_quarters/private_quarters.py

```diff
@@ -0,0 +1,289 @@
+        """
+        Execute purchase weekly roses from shop routine
+        Must have 24K+, try next day if low
+        """
+        # Enter shop
+        self.ui_click(
+            click_button=PRIVATE_QUARTERS_SHOP_ENTER,
+            check_button=PRIVATE_QUARTERS_SHOP_CHECK,
+            appear_button=page_private_quarters.check_button,
+            offset=(20, 20),
+            skip_first_screenshot=False
+        )
+
+        # Roses available for purchase?
+        # Exit shop if not
+        if not self.appear(PRIVATE_QUARTERS_SHOP_WEEKLY_ROSES_CHECK, offset=(20, 20)):
+            logger.info('No more weekly roses to purchase, exit subtask')
+            self._pq_shop_exit()
+            return
+
+        # Read coins, exit if < 24000 (total price for all roses)
+        # Try again next day if low
+        currency = OCR_SHOP_GOLD_COINS.ocr(self.device.image)
+        if currency < 24000:
+            logger.warn(f'Have: {currency}, Need: 24000. Try again next day')
+            return
+        logger.info('Purchasing all available weekly roses')
+
+        # Execute purchase operation
+        skip_first_screenshot=True
+        while 1:
+            if skip_first_screenshot:
+                skip_first_screenshot = False
+            else:
+                self.device.screenshot()
+
+            # End
+            if self.appear(PRIVATE_QUARTERS_SHOP_WEEKLY_ROSES_GET, offset=(20, 20), interval=1):
+                self.device.click(PRIVATE_QUARTERS_SHOP_BACK)
+                break
+
+            if self.appear_then_click(PRIVATE_QUARTERS_SHOP_WEEKLY_ROSES_CHECK, offset=(20, 20), interval=1):
+                continue
+            if self.appear_then_click(PRIVATE_QUARTERS_SHOP_AMOUNT_MAX, offset=(20, 20), interval=1):
+                continue
+            if self.appear_then_click(PRIVATE_QUARTERS_SHOP_CONFIRM_AMOUNT, offset=(20, 20), interval=1):
+                continue
+
+        # Exit shop
+        self._pq_shop_exit()
+
+
+    def pq_interact(self):
+        """
+        Execute target interact routine
+        """
+        click_iteration = 0
+        click_timer = Timer(1.5, count=3).start()
+        skip_first_screenshot = True
+        while 1:
+            if skip_first_screenshot:
+                skip_first_screenshot = False
+            else:
+                self.device.screenshot()
+
+            # End
+            if click_iteration > 2:
+                break
+
+            if click_timer.reached():
+                self.device.click(PRIVATE_QUARTERS_ROOM_TARGET_CLICK_AREA)
+                click_timer.reset()
+            if self.appear_then_click(PRIVATE_QUARTERS_INTERACT, offset=(20, 20), interval=1):
+                click_timer.reset()
+            if self.appear(PRIVATE_QUARTERS_INTERACT_CHECK, offset=(20, 20), interval=1):
+                self.device.click(PRIVATE_QUARTERS_ROOM_BACK)
+                click_timer.reset()
+                click_iteration += 1
+
+        self._pq_goto_room_exit()
```

Comment:
End condition must be image detection

Reply:
would combination of both be acceptable, I think this might work
`if click_iteration > 2 and self.appear(PRIVATE_QUARTERS_INTERACT)`

to complete the task, needs cycle in this loop at least 3 times

Reply:
done in 3eb37033a17bb9af73fb16d984e61ca089e2f6a3

hopefully acceptable, ill keep this comment unresolved if unsatisfactory